### ISSUE TITLE
[14.0 -> 16.0] SQL views removed in Odoo

### ISSUE THREAD

**Issue by remi-filament** [OCA/OpenUpgrade#4313]
## Module
account

## Describe the bug
After running migration from version 15.0 to 16.0, I tried deleting obsolete columns and tables from 16.0 DB with database_cleanup module and it reports that some cannot be deleted since still in use in account_invoice_report view.

I then checked that account_invoice_report view and it still holds the query from old version

## To Reproduce
**Affected versions**: 16.0 (actually I see that view query from v14 is also not updated when migration to v15)

1. Get a v15 db with account module installed
2. Migrate to v16
3. Check your account_invoice_report view query

**Expected behavior**
View should be recreated by Odoo if code changes, if not, it should be done by OpenUpgrade script.

**Additional context**
```
v16_mig=# \d+ account_invoice_report 
                                 View "public.account_invoice_report"
         Column          |       Type        | Collation | Nullable | Default | Storage  | Description 
-------------------------+-------------------+-----------+----------+---------+----------+-------------
 id                      | integer           |           |          |         | plain    | 
 move_id                 | integer           |           |          |         | plain    | 
 product_id              | integer           |           |          |         | plain    | 
 account_id              | integer           |           |          |         | plain    | 
 analytic_account_id     | integer           |           |          |         | plain    | 
 journal_id              | integer           |           |          |         | plain    | 
 company_id              | integer           |           |          |         | plain    | 
 currency_id             | integer           |           |          |         | plain    | 
 commercial_partner_id   | integer           |           |          |         | plain    | 
 name                    | character varying |           |          |         | extended | 
 state                   | character varying |           |          |         | extended | 
 type                    | character varying |           |          |         | extended | 
 partner_id              | integer           |           |          |         | plain    | 
 invoice_user_id         | integer           |           |          |         | plain    | 
 fiscal_position_id      | integer           |           |          |         | plain    | 
 invoice_payment_state   | character varying |           |          |         | extended | 
 invoice_date            | date              |           |          |         | plain    | 
 invoice_date_due        | date              |           |          |         | plain    | 
 invoice_payment_term_id | integer           |           |          |         | plain    | 
 invoice_partner_bank_id | integer           |           |          |         | plain    | 
 residual                | numeric           |           |          |         | main     | 
 amount_total            | numeric           |           |          |         | main     | 
 product_uom_id          | integer           |           |          |         | plain    | 
 product_categ_id        | integer           |           |          |         | plain    | 
 quantity                | numeric           |           |          |         | main     | 
 price_subtotal          | numeric           |           |          |         | main     | 
 price_average           | numeric           |           |          |         | main     | 
 country_id              | integer           |           |          |         | plain    | 
 nbr_lines               | integer           |           |          |         | plain    | 
 team_id                 | integer           |           |          |         | plain    | 
View definition:
 SELECT line.id,
    line.move_id,
    line.product_id,
    line.account_id,
    line.analytic_account_id,
    line.journal_id,
    line.company_id,
    line.company_currency_id AS currency_id,
    line.partner_id AS commercial_partner_id,
    move.name,
    move.state,
    move.move_type AS type,
    move.partner_id,
    move.invoice_user_id,
    move.fiscal_position_id,
    move.payment_state AS invoice_payment_state,
    move.invoice_date,
    move.invoice_date_due,
    move.invoice_payment_term_id,
    move.partner_bank_id AS invoice_partner_bank_id,
    (- line.balance) * (move.amount_residual_signed / NULLIF(move.amount_total_signed, 0.0)) * (line.price_total / NULLIF(line.price_subtotal, 0.0)) AS residual,
    (- line.balance) * (line.price_total / NULLIF(line.price_subtotal, 0.0)) AS amount_total,
    uom_template.id AS product_uom_id,
    template.categ_id AS product_categ_id,
    line.quantity / NULLIF(COALESCE(uom_line.factor, 1::numeric) / COALESCE(uom_template.factor, 1::numeric), 0.0) *
        CASE
            WHEN move.move_type::text = ANY (ARRAY['in_invoice'::character varying::text, 'out_refund'::character varying::text, 'in_receipt'::character varying::text]) THEN '-1'::integer
            ELSE 1
        END::numeric AS quantity,
    - line.balance AS price_subtotal,
    - COALESCE(line.balance / NULLIF(line.quantity, 0.0) / NULLIF(COALESCE(uom_line.factor, 1::numeric), 0.0) / NULLIF(COALESCE(uom_template.factor, 1::numeric), 0.0), 0.0) AS price_average,
    COALESCE(partner.country_id, commercial_partner.country_id) AS country_id,
    1 AS nbr_lines,
    move.team_id
   FROM account_move_line line
     LEFT JOIN res_partner partner ON partner.id = line.partner_id
     LEFT JOIN product_product product ON product.id = line.product_id
     LEFT JOIN account_account account ON account.id = line.account_id
     LEFT JOIN account_account_type user_type ON user_type.id = account.user_type_id
     LEFT JOIN product_template template ON template.id = product.product_tmpl_id
     LEFT JOIN uom_uom uom_line ON uom_line.id = line.product_uom_id
     LEFT JOIN uom_uom uom_template ON uom_template.id = template.uom_id
     JOIN account_move move ON move.id = line.move_id
     LEFT JOIN res_partner commercial_partner ON commercial_partner.id = move.commercial_partner_id
  WHERE (move.move_type::text = ANY (ARRAY['out_invoice'::character varying::text, 'out_refund'::character varying::text, 'in_invoice'::character varying::text, 'in_refund'::character varying::text, 'out_receipt'::character varying::text, 'in_receipt'::character varying::text])) AND line.account_id IS NOT NULL AND NOT line.exclude_from_invoice_tab
  GROUP BY line.id, line.move_id, line.product_id, line.account_id, line.analytic_account_id, line.journal_id, line.company_id, line.currency_id, line.partner_id, move.name, move.state, move.move_type, move.amount_residual_signed, move.amount_total_signed, move.partner_id, move.invoice_user_id, move.fiscal_position_id, move.payment_state, move.invoice_date, move.invoice_date_due, move.invoice_payment_term_id, move.partner_bank_id, uom_template.id, uom_line.factor, template.categ_id, (COALESCE(partner.country_id, commercial_partner.country_id)), move.team_id;
```

**Comment by remi-filament** [comment]
I tried deleting the view manually from postgres and Odoo is still working.
I just realized that since v14, Odoo removes these views with the following pull request : https://github.com/odoo/odoo/pull/56080
So this SQL view (and also other like sale_report) are not needed anymore.

I am wondering if these could still be deleted by OpenUpgrade as not needed anymore or if it should be done by database_cleanup afterwards ?

Any insight here would be welcome @pedrobaeza @StefanRijnhart @legalsylvain

**Comment by StefanRijnhart** [comment]
I have to admit this passed me by as well, so thanks for mentioning this! Given that a view does not contain data to be potentially further processed at any point after the OpenUpgrade process, and the fact that it will be very difficult for database_cleanup to know which database view can be marked for removal, I'd say there is a strong case to remove the database views in the OpenUpgrade migration scripts.

**Comment by pedrobaeza** [comment]
Yes, it may be a complement to be put in OpenUpgrade, although not strictly needed for proper operation. Just to avoid confusion when browsing through SQL.

**Comment by remi-filament** [comment]
Thanks for your feedbacks,

For the record, I have made a first analysis about SQL views per version :

Module | SQL view name | v13 | v14 | v15 | v16 | v17
-- | -- | -- | -- | -- | -- | --
account | account_root | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
account | account_invoice_report | CREATE OR REPLACE VIEW | _table_query | _table_query | _table_query | _table_query
crm | crm_activity_report | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
event_sale | event_sale_report |   |   |   | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
fleet | fleet_vehicle_cost_report |   | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
hr_attendance | hr_attendance_report |   |   | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW |  
hr_contract | hr_contract_history |   |   | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
hr_holidays_calendar → hr_holidays (in v14) | hr_leave_report_calendar | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
hr_holidays | hr_leave_employee_type_report |   |   | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
hr_holidays | hr_leave_report | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
hr_skills | hr_employee_skill_report |   |   |   | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
hr_timesheet_attendance | hr_timesheet_attendance_report | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
hr_timesheet + sale_timesheet | timesheets_analysis_report |   |   |   | _table_query + CREATE OR REPLACE VIEW | _table_query + CREATE OR REPLACE VIEW
hr | hr_employee_public | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
im_livechat | im_livechat_report_channel | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
im_livechat | im_livechat_report_operator | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
l10n_in | l10n_in_account_invoice_report | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW |   |  
l10n_in | l10n_in_payment_report | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW |   |  
l10n_in | l10n_in_exempted_report | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW |   |  
l10n_in | l10n_in_product_hsn_report | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW |   |  
lunch | lunch_cashmove_report | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
lunch | lunch_product_report | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW |   |   |  
mass_mailing | mailing_trace_report | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
membership | report_membership | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
point_of_sale | report_pos_order | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
project | report_project_task_user | CREATE VIEW | CREATE VIEW | CREATE VIEW | CREATE VIEW | CREATE VIEW
project | project_task_burndown_chart_report |   |   | CREATE OR REPLACE VIEW |   |  
purchase | purchase_bill_union | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
purchase | purchase_report | CREATE OR REPLACE VIEW | _table_query | _table_query | _table_query | _table_query
purchase_stock | vendor_delay_report |   | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
report → sale (v14) | report_all_channels_sales | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW |   |  
sale | sale_report | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | _table_query | _table_query
sale_timesheet | project_profitability_report | CREATE VIEW | CREATE VIEW | CREATE VIEW |   |  
stock | report_stock_forecast | CREATE OR REPLACE VIEW |   |   |   |  
stock | report_stock_quantity | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW | CREATE OR REPLACE VIEW
website_crm_partner_assign | crm_partner_report_assign | CREATE OR REPLACE VIEW | _table_query | _table_query | _table_query | _table_query
website_event_questions | event_question_report | CREATE VIEW |   |   |   |  

SQL view added should be properly handled in each version, the issue is related to the ones which were removed.
If I would sum up what should be removed per version : 
- v14 : 
  - account_invoice_report (use _table_query now)
  - purchase_report (use _table_query now)
  - report_stock_forecast (removed from code)
  - crm_partner_report_assign (use _table_query now)
  - event_question_report (removed from code)
- v15 : 
  - lunch_product_report (removed from code)
- v16 : 
  - l10n_in_account_invoice_report (removed from code)
  - l10n_in_payment_report (removed from code)
  - l10n_in_exempted_report (removed from code)
  - l10n_in_product_hsn_report (removed from code)
  - project_task_burndown_chart_report (removed from code)
  - report_all_channels_sales (removed from code)
  - sale_report (use _table_query now)
  - project_profitability_report (removed from code)
- v17 : 
  - hr_attendance_report (removed from code)

Now, regarding way forward : should views be removed in each module script or can these be removed globally (in apriory.py for instance) ? If you prefer in each module should we have 1 PR per module per version ?

**Comment by pedrobaeza** [comment]
Thanks for the deep analysis. It should be done on each module scripts. You can do only one PR for all for not increasing too much time in this. Pointing to the issue and using `[OU-IMP] *: remove obsolete SQL views` or similar commit message seems enough.

**Comment by remi-filament** [comment]
I have just created PR for v14, and sale module for v16. I have added comment in project / sale_timesheet open PRs for v16.
For v15, lunch module has no script (yet ?) so I did not do anything.
Same for l10n_in in v16 and v17 module obviously

Maybe we should update the module used for analysis so that it can detect these cases as well and list them in upgrade_analysis.txt files ? (this I do not know how to do though...)
